clients/src/main/java/org/apache/kafka/common/network/Selector.java

```diff
@@ -529,7 +561,11 @@ else if (e instanceof AuthenticationException) // will be logged later as error
                     log.debug("Connection with {} disconnected due to authentication exception", desc, e);
                 else
                     log.warn("Unexpected error from {}; closing connection", desc, e);
-                close(channel, sendFailed ? CloseMode.NOTIFY_ONLY : CloseMode.GRACEFUL);
+                CloseMode closeMode = sendFailed ? CloseMode.NOTIFY_ONLY : CloseMode.GRACEFUL;
```

Comment:
If `sendFailed==true`, then we there is no point in delaying close, we should just close the channel.

Reply:
It seems `sendFailed` and `prepareFailed` cannot both be `true` at the same time. I think when `prepareFailed` is `true`, we can always close with `CloseMode.GRACEFUL` - we probably won't have any staged receives either way? In any case, I refactored this so that we always close with `CloseMode.GRACEFUL` for delayed authentication close.